### PR TITLE
Add RepositoryStockValueZipExport service and connect export consumption button [SCI-9023]

### PR DESCRIPTION
**Pull request by wandji20**
Jira ticket: [SCI-9023](https://scinote.atlassian.net/browse/SCI-9023)

### What was done
_Add RepositoryStockValueZipExport service and connect export consumption button_

#### ToDo:
_N/A_

### Note:
_N/A_

[SCI-9023]: https://scinote.atlassian.net/browse/SCI-9023?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/assets/javascripts/repositories/repository_datatable.js [hunk 1]
```diff
@@ -438,6 +437,33 @@ var RepositoryDatatable = (function(global) {
   }
 
   function initExportActions() {
+    // RepositoryStockValue Export Action
+    $(document).on('click', '#exportStockConsumptionButton', function(e) {
+      e.preventDefault();
+      e.stopPropagation();
+      let rowIds = [];
```
Comment (okriuchykhin) [inline app/assets/javascripts/repositories/repository_datatable.js:444]: Why don't you simply use rowsSelected directly?

File: app/controllers/repositories_controller.rb [hunk 2]
```diff
@@ -356,6 +356,15 @@ def export_repositories
     end
   end
 
+  def export_repository_stock_items
+    if params[:row_ids]
+      RepositoryStockLedgerZipExport.generate_zip(params[:row_ids], current_user)
```
Comment (okriuchykhin) [inline app/controllers/repositories_controller.rb:361]: how do you check that rows belong to specific repository and user can export them?

File: app/services/repository_stock_ledger_zip_export.rb [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                                     repository_stock_value: {
+                                       repository_ledger_records: [
+                                         { repository_stock_value: :repository_stock_unit_item },
+                                         :reference,
+                                         :user
+                                       ]
+                                     }
+                                   )
+    # rubocop:disable Metrics/BlockLength
+    repository_rows.each do |repository_row|
```
Comment (okriuchykhin) [inline app/services/repository_stock_ledger_zip_export.rb:60]: use find_each

File: app/services/repository_stock_ledger_zip_export.rb [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      previous_balance = nil
+      next unless repository_row.repository_stock_value
```
Comment (okriuchykhin) [inline app/services/repository_stock_ledger_zip_export.rb:62]: use INNER JOIN for this

File: app/services/repository_stock_ledger_zip_export.rb [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+      repository_row.repository_stock_value.repository_ledger_records.order(:created_at).each do |record|
+        action_type = previous_balance && record.balance < previous_balance ? 'reduced' : 'added'
+        consumption_type = record.reference_type == 'MyModuleRepositoryRow' ? 'Task' : 'Inventory'
+
+        if action_type == 'reduced'
+          consumed_amount = record.amount.abs
+          consumed_amount_unit = record.unit || record.repository_stock_value.repository_stock_unit_item.data
```
Comment (okriuchykhin) [inline app/services/repository_stock_ledger_zip_export.rb:70]: load units only from ledger records

File: app/services/repository_stock_ledger_zip_export.rb [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        else
+          added_amount = record.amount
+          added_amount_unit = record.unit || record.repository_stock_value.repository_stock_unit_item.data
```
Comment (okriuchykhin) [inline app/services/repository_stock_ledger_zip_export.rb:73]: same

File: app/services/repository_stock_ledger_zip_export.rb [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        end
+
+        row_data = [
+          consumption_type,
+          repository_row.name,
+          "IT#{repository_row.id}",
```
Comment (okriuchykhin) [inline app/services/repository_stock_ledger_zip_export.rb:79]: repository_row.code

File: app/services/repository_stock_ledger_zip_export.rb [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+          consumed_amount,
+          consumed_amount_unit,
+          added_amount,
+          added_amount_unit,
+          record.user.full_name,
+          record.created_at.strftime(record.user.date_format),
+          record.balance,
+          record.repository_stock_value.repository_stock_unit_item.data
```
Comment (okriuchykhin) [inline app/services/repository_stock_ledger_zip_export.rb:87]: why it is here?

File: app/services/repository_stock_ledger_zip_export.rb [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        ]
+        breadcrumbs_data = Array.new(5, '')
+
+        if consumption_type == 'Task'
+          breadcrumbs_data = [
+            record.reference.my_module.project.team.name,
+            record.reference.my_module.project.name,
+            record.reference.my_module.experiment.name,
+            record.reference.my_module.name,
+            "TA#{record.reference.my_module.id}"
```
Comment (okriuchykhin) [inline app/services/repository_stock_ledger_zip_export.rb:97]: my_module.code

File: app/views/repositories/show.html.erb [hunk 10]
```diff
@@ -59,7 +59,10 @@
              }
   %>
 
-  <div id="actionToolbar" data-behaviour="vue">
+  <div
+    id="actionToolbar" 
+    data-behaviour="vue"
```
Comment (okriuchykhin) [inline app/views/repositories/show.html.erb:64]: why did you format like this?

Comment (wandji20) [reply app/views/repositories/show.html.erb]: Just forgot to remove after reverting previous changes.

File: app/services/repository_stock_ledger_zip_export.rb [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        csv << row
+      end
+    end
+  end
+
+  def self.generate_data(row_ids)
+    data = []
+    repository_rows =
+      RepositoryRow.where(id: row_ids)
+                   .joins(repository_stock_value: [{ repository_ledger_records: [:user] }])
```
Comment (okriuchykhin) [inline app/services/repository_stock_ledger_zip_export.rb:51]: joins(repository_stock_value: { repository_ledger_records: :user })

File: app/services/repository_stock_ledger_zip_export.rb [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+      rows.each do |row|
+        csv << row
+      end
+    end
+  end
+
+  def self.generate_data(row_ids)
+    data = []
+    repository_rows =
+      RepositoryRow.where(id: row_ids)
```
Comment (okriuchykhin) [inline app/services/repository_stock_ledger_zip_export.rb:50]: I think you need to load RepositoryLedgerRecord

File: app/services/repository_stock_ledger_zip_export.rb [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+      rows.each do |row|
+        csv << row
+      end
+    end
+  end
+
+  def self.generate_data(row_ids)
+    data = []
+    repository_ledger_records =
+      RepositoryLedgerRecord.joins([{ repository_stock_value: { repository_cell: :repository_row } }, :user])
```
Comment (okriuchykhin) [inline app/services/repository_stock_ledger_zip_export.rb:50]: joins(:user, repository_stock_value: { repository_cell: :repository_row })

File: app/services/repository_stock_ledger_zip_export.rb [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+                            .where("repository_rows.id IN (#{row_ids.split.join(',')})")
```
Comment (okriuchykhin) [inline app/services/repository_stock_ledger_zip_export.rb:51]: where(repository_row: { id: row_ids })

File: app/services/repository_stock_ledger_zip_export.rb [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+                              repository_ledger_records.reference_id = my_module_repository_rows.id')
+                            .joins('LEFT OUTER JOIN my_modules ON
+                              my_modules.id = my_module_repository_rows.my_module_id')
+                            .joins('LEFT OUTER JOIN experiments ON experiments.id = my_modules.experiment_id')
+                            .joins('LEFT OUTER JOIN projects ON projects.id = experiments.project_id')
+                            .joins('LEFT OUTER JOIN teams ON teams.id = projects.team_id')
+                            .order('row_created_at, repository_ledger_records.created_at')
+                            .select('repository_ledger_records.*,
+                              repository_rows.created_at AS row_created_at,
+                              repository_rows.id AS repository_id,
```
Comment (okriuchykhin) [inline app/services/repository_stock_ledger_zip_export.rb:64]: should be repository_row_id

File: app/services/repository_stock_ledger_zip_export.rb [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+        consumed_amount_unit = record.unit
+      else
+        added_amount = record.amount
+        added_amount_unit = record.unit
+      end
+
+      row_data = [
+        consumption_type,
+        record.repository_name,
+        "IT#{record.repository_id}",
```
Comment (okriuchykhin) [inline app/services/repository_stock_ledger_zip_export.rb:96]: record.repository_row.code

File: app/services/repository_stock_ledger_zip_export.rb [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+        consumed_amount_unit,
+        added_amount,
+        added_amount_unit,
+        record.full_name,
+        record.created_at.strftime(JSON.parse(record.user_settings)['date_format']),
+        record.team_name,
+        record.project_name,
+        record.experiment_name,
+        record.module_name,
+        "TA#{record.module_id}",
```
Comment (okriuchykhin) [inline app/services/repository_stock_ledger_zip_export.rb:107]: "#{MyModule::ID_PREFIX}#{record.module_id}"

**Review by okriuchykhin — APPROVED** [review]

**Pull request merged**